components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,316 @@
+use std::{
+    char::{from_u32, MAX},
+    convert::TryFrom,
+    ops::RangeBounds,
+    slice::Iter,
+};
+
+use super::USetError;
+use crate::utils::{deconstruct_range, is_valid};
+/// Represents the end code point of the Basic Multilingual Plane range, starting from code point 0 , inclusive
+const BMP_MAX: u32 = 0xFFFF;
+
+/// UnicodeSet membership wrapper
+///
+/// Provides exposure to membership functions and constructors from serialized UnicodeSets
+/// and predefined ranges.
+/// Implements an [inversion list.](https://en.wikipedia.org/wiki/Inversion_list)
+#[derive(Debug, PartialEq)]
+pub struct UnicodeSet {
+    // If we wanted to use an array to keep the memory on the stack, there is an unsafe nightly feature
+    // https://doc.rust-lang.org/nightly/core/array/trait.FixedSizeArray.html
+    // Allows for traits of fixed size arrays
+    inv_list: Vec<u32>,
+}
+
+impl TryFrom<Vec<u32>> for UnicodeSet {
```

Comment:
Consider making this an explicit factory method on `UnicodeSet`, e.g.
```
pub fn from_range_endpoints(endpoints: Vec<u32>) -> Result<Self, Self::Error>;
```

The standard traits `TryFrom` and `From` are usually implemented when the input type is unambiguously convertible to the output. But a vector of `u32`s could just as easily be a collection of the individual code points within the set. Clients of the library should not need to know about internal implementation details to figure out what a standard trait implementation might do.

Reply:
I'm torn on this. On one hand, I agree with @kpozin rationale, on the other you can apply it to `FromStr` - a `&str` is any string, and that's why we may fail at parsing it in `FromStr` which is just a specialization of `TryFrom<&str>`.

@Manishearth - do you have any thoughts on this?

One thing that I noticed is that it does consume the argument, so it does convert input into output without allocations, just with verification.
We'll want a similar thing, without checking, for macros later, and it'll likely be some `from_endpoints_vec_unchecked(endpoints: Vec<u32>) -> Self` so maybe `try_from_endpoints_vec` is the right name here?

Overall, I'm not personally concerned with `TryFrom` here, but want to run it via @Manishearth based on this comment.

Reply:
I think `from_range_endpoints` is clearer here, I wasn't sure what this meant here. Also, it should probably operate on a slice?

Reply:
We could also call it `from_inversion_list` to be more explicit.

Reply:
If it operates on the slice, then it'll allocate, right? If it consumes a `Vec` then it doesn't allocate.

Why do you suggest slice?

Reply:
I'm fine with `from_inversion_list` taking a Vec by value.  I don't feel strongly about the name of the function or whether it is `TryFrom`, but I do think we should move the Vec, not take a slice.

Reply:
Ah if we always move the vec we shoudl take a vec, then.

Reply:
agreed and changed to `from_inversion_list`